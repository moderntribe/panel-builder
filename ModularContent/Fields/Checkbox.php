<?php


namespace ModularContent\Fields;


/**
 * Class Radio
 *
 * @package ModularContent\Fields
 *
 * A group of checkboxes
 */
class Checkbox extends Select {

	const LAYOUT_VERTICAL   = 'vertical';
	const LAYOUT_HORIZONTAL = 'horizontal';
	const LAYOUT_INLINE     = 'inline';

	protected $default      = [];
	protected $layout       = self::LAYOUT_VERTICAL;
	protected $option_width = 12; // Fields are laid out in a 12-column grid. This defines how many columns wide this field should be.

	public function __construct( $args = [] ) {
		$this->validate_layout( $args );
		$this->validate_option_width( $args );

		$this->defaults['layout']       = $this->layout;
		$this->defaults['option_width'] = $this->option_width;
		parent::__construct( $args );
	}

	public function get_blueprint() {
		$blueprint                 = parent::get_blueprint();
		$blueprint['default']      = (object) $this->default;
		$blueprint['layout']       = $this->layout;
		$blueprint['option_width'] = $this->option_width;

		return $blueprint;
	}

	/**
	 * Get valid layouts for this field.
	 *
	 * @return array
	 */
	protected function get_valid_layouts() {
		return [ self::LAYOUT_VERTICAL, self::LAYOUT_HORIZONTAL, self::LAYOUT_INLINE ];
	}

	/**
	 * Ensure that the values for layout are valid.
	 *
	 * @param $args
	 */
	protected function validate_layout( $args ) {
		if ( isset( $args['layout'] ) && ! in_array( $args['layout'], $this->get_valid_layouts() ) ) {
			throw new \InvalidArgumentException( 'Layout argument can only be "vertical", "horizontal", or "inline".' );
		}
	}

	/**
	 * Ensure that the option_width value is valid.
	 *
	 * @param $args
	 */
	protected function validate_option_width( $args ) {
		if ( isset( $args['option_width'] ) && ( $args['option_width'] < 1 || $args['option_width'] > 12 ) ) {
			throw new \InvalidArgumentException( 'option_width argument must be between 1-12.' );
		}
	}

	/**
	 * Massage submitted data before it's saved.
	 *
	 * Do not cast to string, checkboxes get arrays
	 * Do not force default on empty submissions, in
	 * case all checkboxes are purposefully unchecked.
	 *
	 * @param array $data
	 *
	 * @return array
	 */
	public function prepare_data_for_save( $data ) {
		return $data;
	}
} 