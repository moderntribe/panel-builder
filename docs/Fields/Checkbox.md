## Checkbox

A group of checkboxes.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |
| `options` | `array` | An array of options to display to the user |
| `default` | `array` | An array of options you want to be selected by default |
| `option_width` | `int` | Defines how many "columns" this field will take up in relation to its parent. Can be 1-12. |

### Example

```php
$field = new Checkbox( array(
  'label' => __('Pick Options'),
  'name' => 'my-field',
  'description' => __( 'Pick the things that you pick' )
  'options' => [
    'first' => __( 'The First Option' ),
    'second' => __( 'The Second Option' ),
  ],
  'default' => [ 'second' => 1 ],
  'option_width' => 6,
) );
```
