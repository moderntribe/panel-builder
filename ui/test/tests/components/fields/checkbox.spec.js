/* global expect b:true */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Checkbox from 'components/fields/checkbox';
import renderer from 'react-test-renderer';

// test data
const OPTIONS1 = [
	{
		label: 'Standard',
		value: 'standard',
	},
	{
		label: 'Cards',
		value: 'cards',
	},
	{
		label: 'Full',
		value: 'full',
	},
];
const OPTIONS2 = [
	{
		label: 'Orange',
		value: '1',
	},
	{
		label: 'Raspberry',
		value: '2',
	},
];
const DEFAULT1 = {
	cards: 1,
};
const DEFAULT2 = {
	full: 1,
};

describe('Checkbox field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Checkbox description="This is a test description" name="layout" label="Some test label" options={OPTIONS1} default={DEFAULT1} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<Checkbox />);
		expect(wrapper.type()).toEqual('div');
	});

	it('has no checkboxes (without properties).', () => {
		const wrapper = shallow(<Checkbox />);
		expect(wrapper.find('input[type="checkbox"]').length).toEqual(0);
	});

	it('has a label', () => {
		const wrapper = shallow(<Checkbox />);
		expect(wrapper.find('label').length).toEqual(1);
	});

	it('allows us to set and edit props', () => {
		const wrapper = mount(<Checkbox description="This is a test description" name="layout" label="Some test label" options={OPTIONS1} default={DEFAULT1} />);
		expect(wrapper.props().description).toEqual('This is a test description');
		wrapper.setProps({ description: 'Change description for test' });
		expect(wrapper.props().description).toEqual('Change description for test');

		expect(wrapper.props().label).toEqual('Some test label');
		wrapper.setProps({ label: 'Change label for test' });
		expect(wrapper.props().label).toEqual('Change label for test');

		expect(wrapper.props().name).toEqual('layout');
		wrapper.setProps({ name: 'fruit' });
		expect(wrapper.props().name).toEqual('fruit');

		expect(wrapper.props().default).toEqual(DEFAULT1);
		wrapper.setProps({ default: DEFAULT2 });
		expect(wrapper.props().default).toEqual(DEFAULT2);

		expect(wrapper.props().options).toEqual(OPTIONS1);
		wrapper.setProps({ options: OPTIONS2 });
		expect(wrapper.props().options).toEqual(OPTIONS2);
	});

	it('allows us to properly formatted radio options', () => {
		const wrapper = mount(<Checkbox description="This is a test description" name="layout" label="Some test label" options={OPTIONS1} default={DEFAULT1} />);
		expect(wrapper.find('ul').find('li').length).toEqual(3);
		expect(wrapper.find('ul li').first().find('input[type="checkbox"]').length).toEqual(1);
		expect(wrapper.find('ul li').first().find('input[type="checkbox"]').props().value).toEqual('standard');
		expect(wrapper.find('ul li').first().find('label').text()).toEqual('Standard');
	});
});
