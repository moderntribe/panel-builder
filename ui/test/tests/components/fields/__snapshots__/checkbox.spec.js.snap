exports[`Checkbox field renders correctly 1`] = `
<div
  className="field vertical panel-field panel-conditional-field">
  <label
    className="label panel-field-label">
    Some test label
    <span
      className="icon"
      onBlur={undefined}
      onClick={undefined}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined} />
  </label>
  <ul
    className="list">
    <li
      className="item"
      style={Object {}}>
      <label
        className="checkboxLabel">
        <input
          checked={undefined}
          className="checkbox"
          data-depth={0}
          data-field="checkbox"
          data-option-type="multiple"
          name="modular-content-layout[]"
          onChange={[Function]}
          tabIndex={0}
          type="checkbox"
          value="standard" />
        <span />
        Standard
      </label>
    </li>
    <li
      className="item"
      style={Object {}}>
      <label
        className="checkboxLabel">
        <input
          checked={undefined}
          className="checkbox"
          data-depth={0}
          data-field="checkbox"
          data-option-type="multiple"
          name="modular-content-layout[]"
          onChange={[Function]}
          tabIndex={0}
          type="checkbox"
          value="cards" />
        <span />
        Cards
      </label>
    </li>
    <li
      className="item"
      style={Object {}}>
      <label
        className="checkboxLabel">
        <input
          checked={undefined}
          className="checkbox"
          data-depth={0}
          data-field="checkbox"
          data-option-type="multiple"
          name="modular-content-layout[]"
          onChange={[Function]}
          tabIndex={0}
          type="checkbox"
          value="full" />
        <span />
        Full
      </label>
    </li>
  </ul>
</div>
`;
